lib/spack/spack/database.py

```diff
@@ -952,35 +924,24 @@ def _read(self):
         """Re-read Database from the data in the set location.
 
         This does no locking, with one exception: it will automatically
-        migrate an index.yaml to an index.json if possible. This requires
-        taking a write lock.
-
+        try to regenerate a missing DB if local. This requires taking a
+        write lock.
         """
         if os.path.isfile(self._index_path):
-            # Read from JSON file if a JSON database exists
-            self._read_from_file(self._index_path, format='json')
-
-        elif os.path.isfile(self._old_yaml_index_path):
-            if (not self.is_upstream) and os.access(
-                    self._db_dir, os.R_OK | os.W_OK):
-                # if we can write, then read AND write a JSON file.
-                self._read_from_file(self._old_yaml_index_path, format='yaml')
-                with lk.WriteTransaction(self.lock):
-                    self._write(None, None, None)
-            else:
-                # Read chck for a YAML file if we can't find JSON.
-                self._read_from_file(self._old_yaml_index_path, format='yaml')
+            # Read from file if a database exists
+            self._read_from_file(self._index_path)
+            return
 
-        else:
-            if self.is_upstream:
-                raise UpstreamDatabaseLockingError(
-                    "No database index file is present, and upstream"
-                    " databases cannot generate an index file")
-            # The file doesn't exist, try to traverse the directory.
-            # reindex() takes its own write lock, so no lock here.
-            with lk.WriteTransaction(self.lock):
-                self._write(None, None, None)
-            self.reindex(spack.store.layout)
+        if self.is_upstream and not os.path.isfile(self._index_path):
```

Comment:
Based on the above changes, if `os.path.isfile(self._index_path) == True`, then we would have exited the function, so I don't think the extra `not os.path.isfile(self._index_path)` check is required here. IMO it's slightly clearer to keep this inside of the `else` clause, or you could do something like:

```
elif self.is_upstream:
  raise ...

self._write(...)
```

Reply:
Done. Originally, I wrote that redundant  check on purpose to make clear to the reader that both conditions need to be satisfied to raise the exception. After re-reading the exception message I agree that it is clear enough to not cause confusion.